### PR TITLE
Allow time span with steps to get_horizons_coord

### PR DESCRIPTION
**Pull request by dstansby**
Fixes https://github.com/sunpy/sunpy/issues/4693; this is currently a draft. Still need to

- [x] Extract the times from the query
- [x] Add tests
- [x] Add a docstring example

### CODE REVIEW

File: sunpy/coordinates/ephemeris.py [hunk 1]
```diff
@@ -216,8 +216,13 @@ def get_horizons_coord(body, time='now', id_type='majorbody', *, include_velocit
...
         If 'smallbody', search by name for asteroids or comets.
         If 'id', search by ID number.
-    time : {parse_time_types}
-        Time to use in a parse_time-compatible format
+    time : {parse_time_types}, `dict`
+        Time to use in a parse_time-compatible format.
+
+        Alternatively, this can be a dictionary defining a range of times and
+        dates; the range dictionary has to be of the form
+        {{'start':’YYYY-MM-DD [HH:MM:SS]’, 'stop':’YYYY-MM-DD [HH:MM:SS]’, 'step':’n[y|d|m|s]’}}.
```
Comment (dstansby) [inline sunpy/coordinates/ephemeris.py:224]: Just coming back to this now, and thought that instead argument could be a list of `[Time, Time, int]` which we internally convert to the jplhorizons string format. Any thoughts @ayshih?

**Comment by dstansby** [comment]
In case anyone is wondering, the blocker at the moment is the test I have added failing. I haven't looked into it in detail, but at a first (naïve) glance I think it should be passing, so further investigation is needed.

File: sunpy/coordinates/ephemeris.py [hunk 2]
```diff
@@ -274,15 +282,31 @@ def get_horizons_coord(body, time='now', id_type='majorbody', *, include_velocit
...
     """
-    obstime = parse_time(time)
-    array_time = np.reshape(obstime, (-1,))  # Convert to an array, even if scalar
+    if isinstance(time, dict):
+        if set(time.keys()) != set(['start', 'stop', 'step']):
+            raise ValueError('time dictionary must have the keys ["start", "stop", "step"]')
+        epochs = time
+        jpl_fmt = '%Y-%m-%d %H:%M:%S'
+        epochs['start'] = parse_time(epochs['start']).strftime(jpl_fmt)
+        epochs['stop'] = parse_time(epochs['stop']).strftime(jpl_fmt)
```
Comment (ayshih) [inline sunpy/coordinates/ephemeris.py:299]:
```suggestion
        epochs['start'] = parse_time(epochs['start']).tdb.strftime(jpl_fmt)
        epochs['stop'] = parse_time(epochs['stop']).tdb.strftime(jpl_fmt)
```

File: sunpy/coordinates/ephemeris.py [hunk 3]
```diff
@@ -294,13 +318,16 @@ def get_horizons_coord(body, time='now', id_type='majorbody', *, include_velocit
...
 
-    # JPL HORIZONS results are sorted by observation time, so this sorting needs to be undone.
-    # Calling argsort() on an array returns the sequence of indices of the unsorted list to put the
-    # list in order.  Calling argsort() again on the output of argsort() reverses the mapping:
-    # the output is the sequence of indices of the sorted list to put that list back in the
-    # original unsorted order.
-    unsorted_indices = obstime.argsort().argsort()
-    result = result[unsorted_indices]
+    if isinstance(time, dict):
+        obstime = parse_time(result['datetime_jd'], format='jd')
```
Comment (ayshih) [inline sunpy/coordinates/ephemeris.py:322]:
```suggestion
        obstime = parse_time(result['datetime_jd'], format='jd', scale='tdb')
```

File: sunpy/coordinates/tests/test_ephemeris.py [hunk 4]
```diff
@@ -115,6 +115,22 @@ def test_get_horizons_coord_array_time():
...
+    # get_horizons_coord() depends on astroquery
+    pytest.importorskip("astroquery")
+
+    time_dict = {'start': '2013-03-01', 'stop': '2013-03-03', 'step': '1d'}
+    time_ref = Time(['2013-03-01', '2013-03-02', '2013-03-03'])
+
+    e = get_horizons_coord('Geocenter', time_dict)
+    e_ref = get_horizons_coord('Geocenter', time_ref)
+
+    assert_quantity_allclose(e.lon, e_ref.lon)
```
Comment (ayshih) [inline sunpy/coordinates/tests/test_ephemeris.py:129]:
```suggestion
    assert_quantity_allclose(e.lon, e_ref.lon, atol=1e-9*u.deg)
```

**Review by ayshih — CHANGES_REQUESTED** [review]
You aren't properly converting to/from TDB.  This fixes the test.

File: sunpy/coordinates/ephemeris.py [hunk 5]
```diff
@@ -274,15 +282,31 @@ def get_horizons_coord(body, time='now', id_type='majorbody', *, include_velocit
         (-25.16107532, 14.59098438, 3.17667664)
      (d_lon, d_lat, d_radius) in (arcsec / s, arcsec / s, km / s)
         (-0.03306548, 0.00052415, -2.66709222)>
+
+    Query the location of Solar Orbiter at a set of 12 regularly sampled times
+    >>> get_horizons_coord('Solar Orbiter',
+                           time={{'start': '2020-12-01',
+                                  'stop': '2020-12-02',
+                                  'step': '12'}})
```
Comment (ayshih) [inline sunpy/coordinates/ephemeris.py:290]:
```suggestion
    ...                    time={{'start': '2020-12-01',
    ...                           'stop': '2020-12-02',
    ...                           'step': '12'}})
```

**Review by ayshih — COMMENTED** [review]
Can you also add a note to the docstring that makes it clear that any times that are provided as strings are interpreted as UTC time (because of `parse_time()`) as opposed to TDB time (which a user of `astroquery` might expect).

File: sunpy/coordinates/ephemeris.py [hunk 6]
```diff
@@ -274,15 +282,31 @@ def get_horizons_coord(body, time='now', id_type='majorbody', *, include_velocit
         (-25.16107532, 14.59098438, 3.17667664)
      (d_lon, d_lat, d_radius) in (arcsec / s, arcsec / s, km / s)
         (-0.03306548, 0.00052415, -2.66709222)>
+
+    Query the location of Solar Orbiter at a set of 12 regularly sampled times
+    >>> get_horizons_coord('Solar Orbiter',
+    ...                    time={{'start': '2020-12-01',
+    ...                           'stop': '2020-12-02',
+    ...                           'step': '12'}})
```
Comment (ayshih) [inline sunpy/coordinates/ephemeris.py:290]:
```suggestion
    ...                           'step': '12'}})  # doctest: +REMOTE_DATA
```

File: changelog/4698.feature.rst [hunk 7]
```diff
@@ -0,0 +1,3 @@
+`sunpy.coordinates.get_horizons_coord` can now be given a start time, end time,
```
Comment (ayshih) [inline changelog/4698.feature.rst:1]:
```suggestion
:func:`sunpy.coordinates.get_horizons_coord` can now be given a start time, end time,
```

File: sunpy/coordinates/ephemeris.py [hunk 8]
```diff
@@ -216,8 +216,17 @@ def get_horizons_coord(body, time='now', id_type='majorbody', *, include_velocit
...
-    time : {parse_time_types}
-        Time to use in a parse_time-compatible format
+    time : {parse_time_types}, `dict`
+        Time to use in a parse_time-compatible format.
+
+        Alternatively, this can be a dictionary defining a range of times and
+        dates; the range dictionary has to be of the form
+        {{'start': start_time, 'stop': stop_time, 'step':’n[y|d|m|s]’}}.
+        ``start_time`` and ``stop_time`` must be in a parse_time-compatible format,
+        and are intepreted as UTC time. ``step`` must be a string with either a
```
Comment (ayshih) [inline sunpy/coordinates/ephemeris.py:226]:
```suggestion
        and are interpreted as UTC time. ``step`` must be a string with either a
```

**Review by ayshih — APPROVED** [review]
Small stuff, otherwise looks good

File: sunpy/coordinates/ephemeris.py [hunk 9]
```diff
@@ -274,15 +283,32 @@ def get_horizons_coord(body, time='now', id_type='majorbody', *, include_velocit
         (-25.16107532, 14.59098438, 3.17667664)
      (d_lon, d_lat, d_radius) in (arcsec / s, arcsec / s, km / s)
         (-0.03306548, 0.00052415, -2.66709222)>
+
+    Query the location of Solar Orbiter at a set of 12 regularly sampled times
+    >>> get_horizons_coord('Solar Orbiter',
```
Comment (ayshih) [inline sunpy/coordinates/ephemeris.py:288]:
```suggestion
    Query the location of Solar Orbiter at a set of 12 regularly sampled times

    >>> get_horizons_coord('Solar Orbiter',
```

Comment (ayshih) [reply sunpy/coordinates/ephemeris.py:288]: (The example isn't rendering properly in the docs)

**Review by ayshih — APPROVED** [review]

**Review by nabobalis — APPROVED** [review]

**Pull request merged**